stl/inc/deque

```diff
@@ -722,42 +722,40 @@ public:
         _Take_contents(_Right);
     }
 
-private:
-    void _Move_assign(deque& _Right, _Equal_allocators) noexcept {
-        _Tidy();
-        _Pocma(_Getal(), _Right._Getal());
-        _Take_contents(_Right);
-    }
-
-    void _Move_assign(deque& _Right, _Propagate_allocators) {
-        auto& _Al       = _Getal();
-        auto& _Right_al = _Right._Getal();
-        if (_Al == _Right_al) {
-            _Move_assign(_Right, _Equal_allocators{});
-        } else {
-            _Alproxy_ty _Alproxy(_Al);
-            _Alproxy_ty _Right_alproxy(_Right_al);
-            _Container_proxy_ptr12<_Alproxy_ty> _Proxy(_Right_alproxy, _Leave_proxy_unbound{});
-            _Tidy();
-            _Pocma(_Al, _Right_al);
-            _Proxy._Bind(_Alproxy, _STD addressof(_Get_data()));
-            _Take_contents(_Right);
-        }
-    }
-
-    void _Move_assign(deque& _Right, _No_propagate_allocators) {
-        if (_Getal() == _Right._Getal()) {
-            _Move_assign(_Right, _Equal_allocators{});
-        } else {
-            assign(
-                _STD make_move_iterator(_Right._Unchecked_begin()), _STD make_move_iterator(_Right._Unchecked_end()));
-        }
-    }
-
 public:
     deque& operator=(deque&& _Right) noexcept(_Alty_traits::is_always_equal::value) {
         if (this != _STD addressof(_Right)) {
-            _Move_assign(_Right, _Choose_pocma<_Alty>{});
+            constexpr auto _Pocma_val = _Choose_pocma_v<_Alty>;
+            if constexpr (_Pocma_val == _Pocma_values::_Equal_allocators) {
+                _Tidy();
+                _Pocma(_Getal(), _Right._Getal());
+                _Take_contents(_Right);
+            } else if constexpr (_Pocma_val == _Pocma_values::_Propagate_allocators) {
+                auto& _Al       = _Getal();
+                auto& _Right_al = _Right._Getal();
+                if (_Al == _Right_al) {
+                    _Tidy();
+                    _Pocma(_Getal(), _Right._Getal());
+                    _Take_contents(_Right);
+                } else {
+                    _Alproxy_ty _Alproxy(_Al);
+                    _Alproxy_ty _Right_alproxy(_Right_al);
+                    _Container_proxy_ptr12<_Alproxy_ty> _Proxy(_Right_alproxy, _Leave_proxy_unbound{});
+                    _Tidy();
+                    _Pocma(_Al, _Right_al);
+                    _Proxy._Bind(_Alproxy, _STD addressof(_Get_data()));
+                    _Take_contents(_Right);
+                }
+            } else {
+                if (_Getal() == _Right._Getal()) {
+                    _Tidy();
+                    _Pocma(_Getal(), _Right._Getal());
+                    _Take_contents(_Right);
+                } else {
+                    assign(_STD make_move_iterator(_Right._Unchecked_begin()),
+                        _STD make_move_iterator(_Right._Unchecked_end()));
+                }
+            }

```

Comment:
Please have a look at how this has been changed for string an vector

1. We should simply early return in case of self assignment
2.  The default is
```cpp
_Tidy();
_Pocma(_Getal(), _Right._Getal());
_Take_contents(_Right);
```

This shoule be at the end.
3. For `_Pocma_values::_Propagate_allocators` we only need to check for inequality and in that case rebind the proxy
4. For non propagating allocators we should check for inequality and then simply call assign and early return

Reply:
This applies for the other containers too

Reply:
Didn't do that for `xhash` due to post-check

Reply:
Thanks a lot for the fixes